rust-symcrypt/src/ecc/mod.rs

```diff
@@ -0,0 +1,627 @@
+    ///
+    /// This function will return a [`SymCryptError::MemoryAllocationFailure`] if there is not enough memory to allocate the key.
+    pub fn generate_key_pair(
+        curve_type: CurveType,
+        ec_key_usage: EcKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let ec_curve = InnerEcCurve::new(curve_type); // Can fail here due to insufficient memory.
+        unsafe {
+            // SAFETY: FFI calls
+            // Stack allocated since we will do SymCryptEckeyAllocate.
+            let key_ptr = InnerEcKey(symcrypt_sys::SymCryptEckeyAllocate(ec_curve.0));
+            if key_ptr.0.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            match symcrypt_sys::SymCryptEckeySetRandom(ec_key_usage.to_symcrypt_flag(), key_ptr.0) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => {
+                    let key = EcKey {
+                        inner_key: key_ptr,
+                        curve_type: curve_type,
+                        ec_key_usage,
+                        has_private_key: true,
+                    };
+                    Ok(key)
+                }
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `set_key_pair()` sets the private and public key information onto the [`EcKey`].
+    ///
+    /// `curve_type` represents the [`CurveType`] of the key.
+    ///
+    /// `private_key` is a `&[u8]` that represents the private key. This key type must match that of the curve type and will fail with [`SymCryptError::InvalidArgument`] if it does not.
+    ///
+    /// `public_key` is an optional `&[u8]` that represents the public key. If no public key is provided, the public key will be derived from the private key. If a public key is provided,
+    /// it must match the format of the private key and will fail with [`SymCryptError::InvalidArgument`] if it does not.
+    ///
+    /// `ec_key_usage` represents the [`EcKeyUsage`] of the key and indicates if the key will be used for [`EcKeyUsage::EcDsa`] or [`EcKeyUsage::EcDh`], or [`EcKeyUsage::EcDhAndEcDsa`].
+    pub fn set_key_pair(
+        curve_type: CurveType,
+        private_key: &[u8],
+        public_key: Option<&[u8]>,
+        ec_key_usage: EcKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let ec_curve = InnerEcCurve::new(curve_type); // Can fail here due to insufficient memory.
+
+        unsafe {
+            //SAFETY: FFI calls
+            // Stack allocated since we will do SymCryptEckeyAllocate.
+            let key_ptr = InnerEcKey(symcrypt_sys::SymCryptEckeyAllocate(ec_curve.0));
+
+            if key_ptr.0.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            let (public_key_ptr, public_key_count) = match public_key {
```

Comment:
>public_key_count [](http://example.com/codeflow?start=33&length=16)

`public_key_length`? Count makes it sound like it's the number of public keys

Reply:
Changed to `public_key_length`